modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/BusyExecutor.java

```diff
@@ -119,12 +150,40 @@ public boolean busyRun(Runnable r) {
      * Task surrounded with try/catch and if it's complete with any exception - resulting future will return
      *
      * @param r Task to execute.
-     * @return Completable future.
+     * @return Completable future with executed flag in result.
      */
     public CompletableFuture<Boolean> submit(Runnable r) {
+        GridBusyLock lock = busyLock;
+
+        CompletableFuture<Boolean> res = new CompletableFuture<>();
+
+        pool.execute(() -> res.complete(busyRun(r, lock)));
+
+        return res;
+    }
+
+    /**
+     * Submit cancellable task to execute in thread pool.
+     *
+     * @param ct Task to execute.
+     * @return Completable future with executed flag in result.
+     */
+    public CompletableFuture<Boolean> submit(CancellableTask ct) {
+        GridBusyLock lock = busyLock;
+
         CompletableFuture<Boolean> res = new CompletableFuture<>();
+        
+        if (busyRun(() -> cancellableTasks.put(ct, ct), lock)) {
```

Comment:
```suggestion
        if (cancellableTasks.put(ct, ct), lock)) {              
```
Busylock here doesn't prevent race between cancellableTasks.put and cancellableTasks.keySet().forEach(CancellableTask::cancel);